File: backend/node/api/src/application/use-cases/fail-run.ts
import { RunNotFoundError } from "../../domain/errors";
import { inject } from "../../infra/DI/container";
import type { RunRepository } from "../../infra/repository/run";

type Input = {
	runId: string;
	error: string;
};

export class FailRunUseCase {
	@inject("run-repository")
	private declare readonly runRepository: RunRepository;

	async execute(input: Input): Promise<[true, undefined] | [undefined, Error]> {
		const [run, error] = await this.runRepository.getById(input.runId);
		if (error) return [undefined, error];
		if (!run) return [undefined, new RunNotFoundError()];

		const [, failError] = run.fail(input.error);
		if (failError) return [undefined, failError];

		const [, saveError] = await this.runRepository.update(run);
		if (saveError) return [undefined, saveError];

		return [true, undefined];
	}
}

File: backend/go/api/internal/run/use-cases/fail-run.go
package usecases

import (
	"context"

	runPkg "github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/run"
)

type FailRunUseCaseDependencies struct {
	RunRepository *runPkg.RunRepository
}

type FailRunUseCase struct {
	FailRunUseCaseDependencies
}

func NewFailRunUseCase(deps FailRunUseCaseDependencies) FailRunUseCase {
	return FailRunUseCase{
		FailRunUseCaseDependencies: deps,
	}
}

type FailRunInput struct {
	RunId string
	Error string
}

func (u *FailRunUseCase) Execute(input FailRunInput) error {
	run, err := u.RunRepository.GetById(context.Background(), input.RunId)
	if err != nil {
		return err
	}

	if run == nil {
		return runPkg.RunNotFoundError
	}

	err = run.Fail(input.Error)
	if err != nil {
		return err
	}

	err = u.RunRepository.Update(context.Background(), run)
	if err != nil {
		return err
	}

	return nil
}
